classes/persistent/period.php

```diff
@@ -82,4 +82,26 @@ protected function validate_enddate($value) {
 
         return $startdate < $value;
     }
+
+    /**
+     * Finds a valid period for a given date and cpdlogbook instance.
+     * Returns the id of the period if found, and 0 if no period is found.
+     *
+     * @param int $date
+     * @param int $cpdlogbookid
+     * @return int
+     * @throws \coding_exception
+     */
+    public static function get_period_for_date($date, $cpdlogbookid) {
```

Comment:
This is the kind of thing where we should start adding some unit tests. In particular I suspect there might be some off by one bugs if you test the end or start date.

So now is as good as any to start wrangle up some tests, have a read of this:

https://docs.moodle.org/dev/PHPUnit

Unlike many frameworks working with the db in moodle is fine during unit tests and you don't need to mock the db. So just create some test periods in a unit test and then assert which periods come out with certain dates

Reply:
Also you could do all the date matching stuff in a single sql call